PageObjects/ShopPage.cs

```diff
@@ -1,43 +1,139 @@
-using System.Collections.Generic;
-using System.Linq;
-using System.Text;
-using System.Threading.Tasks;
+﻿using Microsoft.Playwright;
+using NUnit.Framework;
+using TechTalk.SpecFlow;
 
 namespace HW18_SpecFlow.PageObjects
 {
     [Binding]
     internal class ShopPage
     {
-        private readonly IPage Page;
-        private readonly string testPageUrl = "https://solartechnology.com.ua/shop";
+        private readonly IPage page;
 
         public ShopPage(IPage page)
         {
-            Page = page;
+            this.page = page;
         }
 
-        #region Test DATA:
-        //Locators:
-        //private readonly string tableLocator = ".ReactTable";
-        #endregion
+        public async Task GoToPageURL(string pageUrl)
+        {
+            await page.GotoAsync(pageUrl);
+
+            try
+            {
+                await page.WaitForURLAsync(pageUrl);
+            }
+            catch (PlaywrightException e)
+            {
+                if (e.Message.Contains("crash"))
+                {
+                    Console.WriteLine("Page crashed: " + e.Message);
+                }
+            }
+        }
+
+        public async Task VerifyH1Visability(string h1)
+        {
+            await Assertions.Expect(page.GetByRole(AriaRole.Heading, new() { Name = h1, Level = 1 })).ToBeVisibleAsync();
+        }
+
+        public async Task ClickOnTabLink(string tabName)
+        {
+            await page.GetByRole(AriaRole.Link, new() { Name = tabName }).ClickAsync();
+        }
+
+        public async Task ClickOnFilterButton()
+        {
+            var filterButtonLocator = "//*[contains(@class, 'filter-button')]";
+            var filterButton = page.Locator(filterButtonLocator);
+            await filterButton.ClickAsync();
+        }
+
+        public async Task VerifyFilterChecked(string filterValue)
+        {
+            var filterCheckboxLocator = $"//span[text()='{filterValue}']";
+            var filter = page.Locator(filterCheckboxLocator);
+
+            await filter.CheckAsync();
+            await Assertions.Expect(filter).ToBeCheckedAsync();
+        }
 
-        #region Page
-        public async Task GoToTestPageURL()
+        [Obsolete]
+        public async Task VerifyFilteredProducts(string filterValue)
         {
-            await Page.GotoAsync(testPageUrl);
+            await page.WaitForNavigationAsync();
+            var productTitleLocator = "//*[@class[starts-with(., 'list-product-title')]]";
+            var allProducts = await page.Locator(productTitleLocator).AllInnerTextsAsync();
+            var productsList = allProducts.ToList();
+            Assert.That(productsList.Count, Is.GreaterThan(0), $"Products by locator {productTitleLocator} not found");
+
+            bool isAllContainFilterValue = productsList.All(product => product.ToLower().Contains(filterValue.ToLower()));
+            Assert.That(isAllContainFilterValue, Is.True, $"Not all Products contains the text {filterValue}");
```

Comment:
This approach is not the best. In case of failure you will see result as
Expected: True
But was: False
"Not all Products contains the text filterValue"
Very inconvenient to debug this. Worst case if test fails "sometimes" and you have no idea what happened even after you re-run it. You have to re-run multiple times and there is no guarantee that you will face exactly the same issue and will be able to fix all issues